formats/protobuf-generator/commonMain/src/kotlinx.serialization.protobuf.generator/Generation.kt

```diff
@@ -0,0 +1,383 @@
+package kotlinx.serialization.protobuf.generator
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.protobuf.ProtoIntegerType
+import kotlinx.serialization.protobuf.ProtoNumber
+import kotlinx.serialization.protobuf.ProtoType
+
+@ExperimentalSerializationApi
+public fun generateProto(
+    descriptors: List<SerialDescriptor>,
+    packageName: String? = null,
+    options: Map<String, String> = emptyMap()
+): String {
+    packageName?.let { if (!it.isProtobufFullIdent) throw IllegalArgumentException("Incorrect protobuf package name '$it'") }
+
+    val customTypes = findCustomTypes(descriptors)
+    return generateSchemeString(customTypes, packageName, options)
+}
+
+private data class CustomTypeDeclaration(val name: String, val descriptor: SerialDescriptor)
+
+private fun findCustomTypes(descriptors: List<SerialDescriptor>): List<CustomTypeDeclaration> {
+    val customTypeDescriptorsBySerialName = linkedMapOf<String, SerialDescriptor>()
+    descriptors.forEach { addCustomTypeWithElements(it, customTypeDescriptorsBySerialName) }
+
+    val customTypesByName = linkedMapOf<String, CustomTypeDeclaration>()
+    customTypeDescriptorsBySerialName.values.forEach {
+        val serialNameAsIdent = makeProtobufIdent(it.serialName.substringAfterLast('.', it.serialName))
+        var nameVariantNumber = 1
+        var name = serialNameAsIdent
+        while (customTypesByName.containsKey(name)) {
+            nameVariantNumber++
+            name = "${serialNameAsIdent}_$nameVariantNumber"
+        }
+        customTypesByName[name] = CustomTypeDeclaration(name, it)
```

Comment:
Have you written the test for this functionality on duplicated names?

Reply:
What kind of test are we talking about?
Now if two descriptors have the same `serialName` they are considered to be the same classes.